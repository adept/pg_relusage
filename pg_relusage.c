/*-------------------------------------------------------------------------
 *
 * pg_relusage is a PostgreSQL extension which allows to log relations that were
 * used by the current statement.
 *
 * This code is based on Bruce Momijan lectures and auto_explain.
 *
 * This program is open source, licensed under the PostgreSQL license.
 * For license terms, see the LICENSE file.
 *
 * Copyright (c) 2022, Dmitry Astapov <dastapov@gmail.com>
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"
#include "executor/executor.h"

#include "utils/lsyscache.h"
#include "nodes/pg_list.h"
#if PG_VERSION_NUM >= 90600
#include "access/parallel.h"
#endif

PG_MODULE_MAGIC;

static int	pg_relusage_log_level = LOG;

static ExecutorStart_hook_type prev_ExecutorStart = NULL;

void		_PG_init(void);
void		_PG_fini(void);

static void pg_relusage_ExecutorStart(QueryDesc *queryDesc, int eflags);

void
_PG_init(void)
{
  //EmitWarningsOnPlaceholders("pg_relusage");

  /* Install hooks only on leader. */
#if PG_VERSION_NUM >= 90600
  if (!IsParallelWorker())
    {
#endif
      prev_ExecutorStart = ExecutorStart_hook;
      ExecutorStart_hook = pg_relusage_ExecutorStart;
#if PG_VERSION_NUM >= 90600
    }
#endif
}

void
_PG_fini(void)
{
#if PG_VERSION_NUM >= 90600
  if (!IsParallelWorker())
    {
#endif
      ExecutorStart_hook = prev_ExecutorStart;
#if PG_VERSION_NUM >= 90600
    }
#endif
}

static void
pg_relusage_ExecutorStart(QueryDesc *queryDesc, int eflags)
{
  ListCell   *lst;
  ListCell   *lst2;
  Oid oid;
  StringInfoData buf;
  bool buf_empty=true;
  char kind;
    
  initStringInfo(&buf);

  foreach(lst, queryDesc -> plannedstmt -> relationOids) {
    bool seen = false;
    oid = lfirst_oid(lst);
    kind = get_rel_relkind(oid);

    // Exclude things with unknown relkind, and ignore indices and TOAST tables
    if (!kind || kind == 'i' || kind == 'I' || kind == 't') continue;
    
    // Oid list is not deduplicated so we check if the same oid was mentioned before
    foreach(lst2, queryDesc -> plannedstmt -> relationOids) {
      if(lst2 == lst)
        break;
      if(oid == lfirst_oid(lst2)) 
        seen = true;
    }
    
    if (!seen) {
      if (buf_empty) {
        buf_empty=false;
      } else {
        appendStringInfoChar(&buf, ',');
      };
      appendStringInfoString(&buf, get_rel_name(oid));
    }
    
  };

  if (!buf_empty) {
    ereport(pg_relusage_log_level, (errmsg("relations used: %s", buf.data), errhidestmt(true)));
  };
  
  pfree(buf.data);
  
  if (prev_ExecutorStart)
    prev_ExecutorStart(queryDesc, eflags);
  else
    standard_ExecutorStart(queryDesc, eflags);
}
