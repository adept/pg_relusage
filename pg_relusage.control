# pg_relusage extension
comment = 'Log all the queries that reference a particular column'
default_version = '0.0.1'
module_pathname = '$libdir/pg_relusage'
