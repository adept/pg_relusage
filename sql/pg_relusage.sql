load 'pg_relusage';
set client_min_messages to LOG;
---------
-- TABLES
---------
create table test(id serial, t text);

-- we detect 'test' in a simple query
\copy (select * from test) to '/dev/null';

-- multiple references reported only once
\copy (select * from test t1 join test t2 using (id)) to '/dev/null';

-- this works with built-in tables as well
\copy (select * from pg_class c1 join pg_class c2 using (relname) limit 1) to '/dev/null';

-- insert, update, delete statements all covered
insert into test(t) values ('test');
update test set t = t||t;
delete from test;

--------
-- VIEWS
--------
create view vw_test as select * from test;

-- we detect 'test' as the only relation used here
\copy (select * from vw_test t1 join test t2 using (id)) to '/dev/null';

-- more complex built-in view
\copy (select * from pg_stats limit 1) to '/dev/null';

-------
-- CTEs
-------
\copy (with stats as (select * from pg_stats limit 1) select * from stats) to '/dev/null';

------------------------
-- TEMP TABLES AND VIEWS
------------------------
create temp table temp_test(t text);
create temp view vw_temp_test as select * from temp_test;
\copy (select * from vw_temp_test) to '/dev/null';
