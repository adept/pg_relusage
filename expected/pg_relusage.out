load 'pg_relusage';
set client_min_messages to LOG;
---------
-- TABLES
---------
create table test(id serial, t text);
-- we detect 'test' in a simple query
\copy (select * from test) to '/dev/null';
LOG:  relations used: test
-- multiple references reported only once
\copy (select * from test t1 join test t2 using (id)) to '/dev/null';
LOG:  relations used: test
-- this works with built-in tables as well
\copy (select * from pg_class c1 join pg_class c2 using (relname) limit 1) to '/dev/null';
LOG:  relations used: pg_class
-- insert, update, delete statements all covered
insert into test(t) values ('test');
LOG:  relations used: test,test_id_seq
update test set t = t||t;
LOG:  relations used: test
delete from test;
LOG:  relations used: test
--------
-- VIEWS
--------
create view vw_test as select * from test;
-- we detect 'test' as the only relation used here
\copy (select * from vw_test t1 join test t2 using (id)) to '/dev/null';
LOG:  relations used: test,vw_test
-- more complex built-in view
\copy (select * from pg_stats limit 1) to '/dev/null';
LOG:  relations used: pg_stats,pg_statistic,pg_class,pg_attribute,pg_namespace
-------
-- CTEs
-------
\copy (with stats as (select * from pg_stats limit 1) select * from stats) to '/dev/null';
LOG:  relations used: pg_stats,pg_statistic,pg_class,pg_attribute,pg_namespace
------------------------
-- TEMP TABLES AND VIEWS
------------------------
create temp table temp_test(t text);
create temp view vw_temp_test as select * from temp_test;
\copy (select * from vw_temp_test) to '/dev/null';
LOG:  relations used: vw_temp_test,temp_test
------
-- GUC
------
set pg_relusage.log_level = 'INFO';
set pg_relusage.rel_kinds = 'r';
\copy (select * from pg_stats limit 1) to '/dev/null';
INFO:  relations used: pg_statistic,pg_class,pg_attribute,pg_namespace
set pg_relusage.log_level = 'NOTICE';
set pg_relusage.rel_kinds = 'v';
\copy (select * from pg_stats limit 1) to '/dev/null';
NOTICE:  relations used: pg_stats
